### PR TITLE
OCPBUGS-36553: VSphere: add vips to machinenetwork

### PR DESCRIPTION
**Pull request by patrickdillon**
Appends the ingress and api vips to the machinenetworks in the cluster infrastructure object, because the cluster network operator checks to ensure that the machinenetwork contains those vips. In some edge cases we have seen valid configurations fail when the VIPs appear to be outside the valid range.

Here are some dumps of what the infrastructure manifest looks like (these were captured during unit tests):

### Spec:

```
    APIServerInternalIPs: ([]v1.IP) (len=1 cap=1) {
     (v1.IP) (len=11) "192.168.1.0"
    },
    IngressIPs: ([]v1.IP) (len=1 cap=1) {
     (v1.IP) (len=13) "192.168.222.4"
    },
    MachineNetworks: ([]v1.CIDR) (len=3 cap=4) {
     (v1.CIDR) (len=11) "10.0.0.0/16",
     (v1.CIDR) (len=14) "192.168.1.0/32",
     (v1.CIDR) (len=16) "192.168.222.4/32"
    }
```

### Status:

```
    APIServerInternalIP: (string) (len=11) "192.168.1.0",
    APIServerInternalIPs: ([]string) (len=1 cap=1) {
     (string) (len=11) "192.168.1.0"
    },
    IngressIP: (string) (len=13) "192.168.222.4",
    IngressIPs: ([]string) (len=1 cap=1) {
     (string) (len=13) "192.168.222.4"
    },
    NodeDNSIP: (string) "",
    LoadBalancer: (*v1.VSpherePlatformLoadBalancer)(<nil>),
    MachineNetworks: ([]v1.CIDR) (len=3 cap=4) {
     (v1.CIDR) (len=11) "10.0.0.0/16",
     (v1.CIDR) (len=14) "192.168.1.0/32",
     (v1.CIDR) (len=16) "192.168.222.4/32"
    }
```

### Spec

```
      APIServerInternalIPs: ([]v1.IP) (len=2 cap=2) {
       (v1.IP) (len=11) "192.168.1.0",
       (v1.IP) (len=7) "fe80::1"
      },
      IngressIPs: ([]v1.IP) (len=2 cap=2) {
       (v1.IP) (len=13) "192.168.222.4",
       (v1.IP) (len=7) "fe80::2"
      },
      MachineNetworks: ([]v1.CIDR) (len=6 cap=8) {
       (v1.CIDR) (len=11) "10.0.0.0/16",
       (v1.CIDR) (len=9) "fe80::/10",
       (v1.CIDR) (len=14) "192.168.1.0/32",
       (v1.CIDR) (len=10) "fe80::1/64",
       (v1.CIDR) (len=16) "192.168.222.4/32",
       (v1.CIDR) (len=10) "fe80::2/64"
      }
```

### Status

```
      APIServerInternalIP: (string) (len=11) "192.168.1.0",
      APIServerInternalIPs: ([]string) (len=2 cap=2) {
       (string) (len=11) "192.168.1.0",
       (string) (len=7) "fe80::1"
      },
      IngressIP: (string) (len=13) "192.168.222.4",
      IngressIPs: ([]string) (len=2 cap=2) {
       (string) (len=13) "192.168.222.4",
       (string) (len=7) "fe80::2"
      },
      NodeDNSIP: (string) "",
      LoadBalancer: (*v1.VSpherePlatformLoadBalancer)(<nil>),
      MachineNetworks: ([]v1.CIDR) (len=6 cap=8) {
       (v1.CIDR) (len=11) "10.0.0.0/16",
       (v1.CIDR) (len=9) "fe80::/10",
       (v1.CIDR) (len=14) "192.168.1.0/32",
       (v1.CIDR) (len=10) "fe80::1/64",
       (v1.CIDR) (len=16) "192.168.222.4/32",
       (v1.CIDR) (len=10) "fe80::2/64"
      }
```
/cc @mkowalski @jcpowermac @rvanderp3

### CODE REVIEW

File: pkg/asset/manifests/vsphere/infrastructure.go [hunk 1]
```diff
@@ -98,3 +101,16 @@ func GetInfraPlatformSpec(ic *installconfig.InstallConfig, clusterID string) *co
...
+	var cidrs []configv1.CIDR
+	for _, vip := range vips {
+		mask := "/32"
+		if strings.Contains(vip, ":") {
+			mask = "/64"
+		}
+		cidrs = append(cidrs, configv1.CIDR(vip+mask))
+	}
+	return cidrs
+}
```
Comment (patrickdillon) [inline pkg/asset/manifests/vsphere/infrastructure.go:116]: How bad is this? I have little idea what I'm doing, so please help!

Comment (mkowalski) [reply pkg/asset/manifests/vsphere/infrastructure.go:116]: For IPv6 let's use `:128` if the desire is to add a single IP address

Comment (mkowalski) [reply pkg/asset/manifests/vsphere/infrastructure.go:116]: Also, the logic of checking if IP address is IPv6 by checking if it contains `:` is quite error-prone, but for what we are doing it it should not matter (i.e. nothing is going to break if you add /32 when you wanted to add /128)

Comment (patrickdillon) [reply pkg/asset/manifests/vsphere/infrastructure.go]: Thanks! Updated.

**Comment by patrickdillon** [comment]
Added a new commit to remove the deprecated APIVIP & IngressVIP fields.

File: pkg/asset/manifests/vsphere/infrastructure.go [hunk 2]
```diff
@@ -98,3 +101,16 @@ func GetInfraPlatformSpec(ic *installconfig.InstallConfig, clusterID string) *co
...
 }
+
+// vipsToCIDRs takes a single ip address and converts it to CIDR notation.
+func vipsToCIDRs(vips []string) []configv1.CIDR {
+	cidrs := make([]configv1.CIDR, len(vips))
+	for i, vip := range vips {
+		mask := "/32"
+		if strings.Count(vip, ":") >= 2 {
+			mask = "/128"
+		}
```
[the remark this answers is not in the document]
Comment (patrickdillon) [reply pkg/asset/manifests/vsphere/infrastructure.go:112]: Yes! Love it. I will update in a bit.

[the remark this answers is not in the document]
Comment (tthvo) [reply pkg/asset/manifests/vsphere/infrastructure.go]: Nice, thanks!

**Review by tthvo — COMMENTED** [review]
/lgtm

**Comment by patrickdillon** [comment]
/cc @vanderp3 @jcpowermac @vr4manta @mkowalski 
PTAL

**Comment by mkowalski** [comment]
/test unit

Looks like a valid failure?

```
 --- FAIL: Test_ensureDNS (10.44s)
    --- FAIL: Test_ensureDNS/valid_dns (10.44s)
        validation_test.go:617: [platform: Invalid value: "api.0a000803.nip.io": lookup api.0a000803.nip.io on 172.30.0.10:53: server misbehaving, platform: Invalid value: "api-int.0a000803.nip.io": lookup api-int.0a000803.nip.io on 172.30.0.10:53: server misbehaving] 
```

File: pkg/asset/manifests/infrastructure_test.go [hunk 3]
```diff
@@ -11,12 +11,23 @@ import (
...
 	azuretypes "github.com/openshift/installer/pkg/types/azure"
 	"github.com/openshift/installer/pkg/types/dns"
 	gcptypes "github.com/openshift/installer/pkg/types/gcp"
 	nonetypes "github.com/openshift/installer/pkg/types/none"
+	vspheretypes "github.com/openshift/installer/pkg/types/vsphere"
+)
+
+const (
+	defaultMachineNetwork = "10.0.0.0/16"
+	ipv6MachineNetwork    = "fe80::/10"
```
Comment (mkowalski) [inline pkg/asset/manifests/infrastructure_test.go:26]: I just noticed, this is actually not a good example. There are multiple places where we specifically forbid using link-local IPv6 range. It may pass the unit test here because those are designed to be simple, but it can stop working any day.

The bad ones are also IPv4-mapped subnets (`::FFFF:0:0/96`) and multicast (`ff00::/8`). I think a reasonable choice can be documentation subnet `2001:db8::/32` because it's not meant to be treated in any special way

Comment (patrickdillon) [reply pkg/asset/manifests/infrastructure_test.go:26]: Thanks Mat. I have a lot to learn about networking & particularly ipv6, so I appreciate your guidance. I will update the unit tests in a bit.

Comment (patrickdillon) [reply pkg/asset/manifests/infrastructure_test.go]: fixed and squashed commits

**Review by tthvo — COMMENTED** [review]
/lgtm
